### PR TITLE
Add to/from arrow for fixed point columns

### LINKED ISSUE #7116: [FEA] Fixed-point support for to_arrow and from_arrow
**Is your feature request related to a problem? Please describe.**
Spark supports fixed-point precision for conversion to/from Arrow

**Describe the solution you'd like**
We need to support to_arrow and from_arrow for fixed-point precision

**Additional context**
This builds upon #3556

### PR DESCRIPTION
**Pull request by shwina**
Closes #7116

### CODE REVIEW

**Comment by codereport** [comment]
rerun tests

File: cpp/src/interop/from_arrow.cu
Comment (jrhemstad) [inline cpp/src/interop/from_arrow.cu:169]: The comment is helpful, but can `2` be replaced with the equivalent equation? Then there wouldn't be a need for the comment.

Comment (codereport) [reply cpp/src/interop/from_arrow.cu:169]: We tried digging around the Arrow API to get the width of the `Arrow::Type:type` width, but gave up trying to find it.

Comment (jrhemstad) [reply cpp/src/interop/from_arrow.cu:169]: Looks like `arrow::Decimal128::bit_width` should give you what you want:

https://github.com/apache/arrow/blob/bdbd1b361f903750918542093545c506de585049/cpp/src/arrow/util/basic_decimal.h#L45

It's defined in the `BasicDecimal128` type from which `Decimal128` inherits:

https://github.com/apache/arrow/blob/bdbd1b361f903750918542093545c506de585049/cpp/src/arrow/util/decimal.h#L50

Comment (codereport) [reply cpp/src/interop/from_arrow.cu:169]: It looks like this was only added [4 months ago](https://github.com/apache/arrow/commit/feabdf071abac12e6dc796c115d1f7567bcac9e2) and it we don't have it in our version of Arrow.

Comment (jrhemstad) [reply cpp/src/interop/from_arrow.cu:169]: I suspect something similar was present in older versions of Arrow, but I think we're upgrading soon anyways, so maybe it doesn't matter.

**Comment by codereport** [comment]
The CI failure:
```
CMakeFiles/cudf.dir/src/interop/to_arrow.cu.o.d
09:43:07 $PREFIX/include/arrow/array/array_nested.h:93:50: error: 'arrow::BaseListArray' is not a type
09:43:07    friend void internal::SetListData<TYPE>(BaseListArray<TYPE>* self,
09:43:07                                                   ^~~~~~~~~~~~~
09:43:07 $PREFIX/include/arrow/array/array_nested.h: In instantiation of 'class arrow::BaseListArray<arrow::ListType>':
09:43:07 $PREFIX/include/arrow/array/array_nested.h:103:63:   required from here
09:43:07 $PREFIX/include/arrow/array/array_nested.h:93:13: error: template-id 'SetListData<arrow::ListType>' for 'void arrow::internal::SetListData(int*, const std::shared_ptr<arrow::ArrayData>&, arrow::Type::type)' does not match any template declaration
09:43:07    friend void internal::SetListData<TYPE>(BaseListArray<TYPE>* self,
09:43:07              ^~~~~~~~
09:43:07 $PREFIX/include/arrow/array/array_nested.h:52:28: note: candidate is: template<class TYPE> void arrow::internal::SetListData(arrow::BaseListArray<TYPE>*, const std::shared_ptr<arrow::ArrayData>&, arrow::Type::type)
09:43:07  template <typename TYPE>
```
is only failing on CUDA 10.1 and it is due to having changed the file from `.cpp` to `.cu`. Jake mentioned that this sounds like an NVCC bug. 

What should we do for now? @kkraus14 @jrhemstad

File: python/cudf/cudf/core/column/column.py
Comment (kkraus14) [inline python/cudf/cudf/core/column/column.py:373]: Is this cast zero copy? Should we be using `result.view(...)` instead?

File: python/cudf/cudf/tests/test_decimal.py
Comment (kkraus14) [inline python/cudf/cudf/tests/test_decimal.py:28]: Link to a JIRA issue if possible

File: python/cudf/cudf/core/column/column.py
Comment (shwina) [reply python/cudf/cudf/core/column/column.py:373]: Amazing! Doing so also lets us get rid of the xfail below.

**Review by kkraus14 — APPROVED** [review]

**Comment by galipremsagar** [comment]
rerun tests

**Review by galipremsagar — APPROVED** [review]

**Comment by jrhemstad** [comment]
> is only failing on CUDA 10.1 and it is due to having changed the file from `.cpp` to `.cu`. Jake mentioned that this sounds like an NVCC bug.
> 
> What should we do for now? @kkraus14 @jrhemstad

@kkraus14 Can this feature wait until 0.20 where we drop 10.1?

**Comment by kkraus14** [comment]
> @kkraus14 Can this feature wait until 0.20 where we drop 10.1?

@sameerz is that okay on your side?

**Review by vuule — CHANGES_REQUESTED** [review]
Looks good, couple of minor suggestions.

**Comment by sameerz** [comment]
> >@kkraus14 Can this feature wait until 0.20 where we drop 10.1?

> @sameerz is that okay on your side?

@kkraus14 Yes we can wait until 0.20.

**Comment by kkraus14** [comment]
Updated target to `branch-0.20`

File: cpp/src/interop/from_arrow.cu
Comment (vuule) [inline cpp/src/interop/from_arrow.cu:221]: I just realized that move is probably not needed here, since `*temp_mask.release()` is a temporary. Not a big deal either way.

**Review by vuule — APPROVED** [review]

File: cpp/CMakeLists.txt
Comment (harrism) [inline cpp/CMakeLists.txt:216]: If it's only decimal code that requires NVCC compilation, suggest keeping everything else in the old .cpp files, and putting only the CUDA/thrust-using functions in `[to|from]_arrrow.cu`. This should be better for compilation time (reduces NVCC bottleneck, improves compilation concurrency)

File: cpp/src/interop/from_arrow.cu
Comment (harrism) [inline cpp/src/interop/from_arrow.cu:211]:
```suggestion
                 gather_map,
```

**Review by harrism — DISMISSED** [review]
Looks good, just a request to isolate the .cu-requiring code.

File: cpp/src/interop/from_arrow.cu
Comment (codereport) [reply cpp/src/interop/from_arrow.cu:211]: Done.

File: cpp/CMakeLists.txt
Comment (codereport) [reply cpp/CMakeLists.txt:216]: @harrism is this something we really want to do? My first attempt was at just isolating the `operator()` specialization for `decimal64` to a `.cu` and then converting the old file back to `.cpp`. This necessitates removing the anonymous namespace that the `dispatch_to_cudf_column` lives in and then moving that struct and all of its member function declarations to the `cudf/detail/interop.hpp` so it can be shared across the `.cu` & `.cpp` (note I don't even have this compiling yet) ...
```cpp
// Now in cudf/detail/interop.hpp
struct dispatch_to_cudf_column {
  template <typename T>
  std::unique_ptr<column> operator()(arrow::Array const& array,
                                     data_type type,
                                     bool skip_mask,
                                     rmm::cuda_stream_view stream,
                                     rmm::mr::device_memory_resource* mr);

  std::unique_ptr<rmm::device_buffer> get_mask_buffer(arrow::Array const& array,
                                                      rmm::cuda_stream_view stream,
                                                      rmm::mr::device_memory_resource* mr);

  std::unique_ptr<column> get_empty_type_column(size_type size);

  std::unique_ptr<column> get_column(arrow::Array const& array,
                                     data_type type,
                                     bool skip_mask,
                                     rmm::cuda_stream_view stream,
                                     rmm::mr::device_memory_resource* mr);
};
```

Comment (harrism) [reply cpp/CMakeLists.txt:216]: It used to be a .cop and worked. What changed? My suggestion is to put whatever changed into the .Cu, but not everything. Sorry on mobile so can't examine the code easily.

Comment (jrhemstad) [reply cpp/CMakeLists.txt:216]: The problem is that "the stuff that changed" is specializations of a type-dispatched functor. Separating that into a .cu is non-trivial. This feels like premature compile time optimization.

Comment (codereport) [reply cpp/CMakeLists.txt:216]: Basically what Jake said. If this was a standalone function or piece of functionality, it would be straight forward. However, the fact the the code lives in an anonymous namespace and is a template specialization of a type-dispatched functor, it makes it very tricky.

**Comment by codereport** [comment]
@harrism Are you able to approve? looks like you completed a review. We can merge once he have the last approval.

**Review by jrhemstad — APPROVED** [review]

**Comment by codereport** [comment]
@gpucibot merge

**Pull request merged**
